staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go

```diff
@@ -361,15 +352,10 @@ func (p *jsonPatcher) applyJSPatch(versionedJS []byte) (patchedJS []byte, retErr
 		}
 		return patchedJS, nil
 	case types.MergePatchType:
-		// sanity check potentially abusive patches
-		// TODO(liggitt): drop this once golang json parser limits stack depth (https://github.com/golang/go/issues/31789)
-		if len(p.patchBytes) > 1024*1024 {
-			v := map[string]interface{}{}
-			if err := json.Unmarshal(p.patchBytes, &v); err != nil {
-				return nil, errors.NewBadRequest(fmt.Sprintf("error decoding patch: %v", err))
-			}
+		v := map[string]interface{}{}
+		if err := json.Unmarshal(p.patchBytes, &v); err != nil {
```

Comment:
this change unmarshals all merge patches twice... the previous code only did so for ones over 1MB... I'd prefer not to regress patch performance as part of removing this workaround.

Reply:
What I want to achieve here is that when users sent patchBytes fail to be unmarshalled for MergePatchType, api-server will return a bad request error same as for JSONPatchType, not simply remove the sanity check.

The previous failing tests try to send bytes over 1MB and check whether get a bad request error which is returned by the sanity check. But for bytes below 1MB, even they cannot be unmarshalled successfully, it will return `fmt.Errorf("Invalid JSON Patch")` instead of bad request. Removing the sanity check will make bytes over 1MB return `fmt.Errorf("Invalid JSON Patch")` because they cannot be unmarshalled successfully as well.

For performance issue, does it make sense to make `jsonpatch.errBadJSONPatch` public and then we can check it here to avoid duplicated unmarshal? 

https://github.com/evanphx/json-patch/blob/147e2b447e4274fdaacdf8d02e08e87933d3120e/merge.go#L92

Reply:
> For performance issue, does it make sense to make jsonpatch.errBadJSONPatch public and then we can check it here to avoid duplicated unmarshal?

Some way of understanding whether the error came from the doc, the patch, or the application of the patch to the doc would make sense to me, so we could return a BadRequest if the patch itself was bad.